src/base/bittorrent/session.cpp

```diff
@@ -671,124 +637,130 @@ QString Session::torrentTempPath(const TorrentInfo &torrentInfo) const
     return tempPath();
 }
 
-bool Session::isValidCategoryName(const QString &name)
-{
-    QRegExp re(R"(^([^\\\/]|[^\\\/]([^\\\/]|\/(?=[^\/]))*[^\\\/])$)");
-    if (!name.isEmpty() && (re.indexIn(name) != 0)) {
-        qDebug() << "Incorrect category name:" << name;
-        return false;
-    }
-
-    return true;
-}
-
-QStringList Session::expandCategory(const QString &category)
+QStringList Session::expandCategoryName(const QString &categoryName)
 {
     QStringList result;
-    if (!isValidCategoryName(category))
-        return result;
 
     int index = 0;
-    while ((index = category.indexOf('/', index)) >= 0) {
-        result << category.left(index);
+    while ((index = categoryName.indexOf('/', index)) >= 0) {
+        result << categoryName.left(index);
         ++index;
     }
-    result << category;
+    result << categoryName;
 
     return result;
 }
 
-QStringList Session::categories() const
+const CategoryMap Session::categories() const
 {
-    return m_categories.keys();
+    return m_categories;
 }
 
 QString Session::categorySavePath(const QString &categoryName) const
 {
-    QString basePath = m_defaultSavePath;
-    if (categoryName.isEmpty()) return basePath;
+    if (categoryName.isEmpty())
+        return defaultSavePath();
 
-    QString path = m_categories.value(categoryName);
-    if (path.isEmpty()) // use implicit save path
-        path = Utils::Fs::toValidFileSystemName(categoryName, true);
+    QString path {Utils::Fs::toValidFileSystemName(categoryName, true)};
+    TorrentCategory *const category {findCategory(categoryName)};
+    if (category && !category->savePath().isEmpty()) // use explicit save path
+        path = category->savePath();
 
     if (!QDir::isAbsolutePath(path))
-        path.prepend(basePath);
+        path = QDir(defaultSavePath()).absoluteFilePath(path);
 
     return normalizeSavePath(path);
 }
 
-bool Session::addCategory(const QString &name, const QString &savePath)
+TorrentCategory *Session::findCategory(const QString &categoryName) const
 {
-    if (name.isEmpty()) return false;
-    if (!isValidCategoryName(name) || m_categories.contains(name))
-        return false;
+    return m_categories.value(TorrentCategory::cleanCategoryName(categoryName));
+}
+
+TorrentCategory *Session::getCategory(const QString &categoryName)
+{
+    const QString validCategoryName {TorrentCategory::cleanCategoryName(categoryName)};
+    TorrentCategory *category = m_categories.value(validCategoryName);
+    if (category) return category;
+
+    return createCategory_impl(validCategoryName);
+}
+
+TorrentCategory *Session::createCategory(const QString &categoryName)
+{
+    return createCategory_impl(TorrentCategory::cleanCategoryName(categoryName));
+}
+
+TorrentCategory *Session::createCategory_impl(const QString &validCategoryName)
+{
+    if (validCategoryName.isEmpty())
+        return nullptr;
+
+    if (m_categories.contains(validCategoryName))
+        return nullptr;
```

Comment:
Someone from the outside world may try to create category with empty name. It's expected here. We just don't allow it.

Reply:
It's a private function. If someone from outside calls it with an unchecked parameter value, you are in trouble.

Reply:
That may be so, but what am I supposed to do with configuration files? They are also outside world! The user can go in there and change anything...
Now I have 3 entry points from outside which call this private method: createCategory(), getCategory() and loadCategories(). Do you believe I should have checks for empty string in all these places and assertion here?

Reply:
> Now I have 3 entry points from outside that call this private method: createCategory(), getCategory() and loadCategories(). Do you believe I should have checks for empty string in all these places and assertion here?

Good point! Clearly shows the design mistake. Temptation to do parameters checking in a private function shows absence of a strict contract and usually indicates a mistake in the subsystem interface. In this case, `TorrentCategory` is clearly an underdeveloped class (does not really differ from `QPair<QString,QString>`).

Reply:
>Good point!

Do you have anything here apart from gloating?
>Temptation to do parameters checking in a private function shows absence of a strict contract and usually indicates a mistake in the subsystem interface.

Agree. But where is this mistake?

Reply:
Sorry, I did not mean to be offensive. IMO, the mistake is in the fact that `TorrentCategory` does not have a true constructor, in mapping string key to a `TorrentCategory` object containing the same key.

Reply:
>TorrentCategory does not have a true constructor, in mapping string key to a TorrentCategory object containing the same key.

Please rephrase it or explain in detail. What is the constructor supposed to be? If you mean the constructor, which clean the category name inside it, then it is undoubtedly the most correct (this is similar to QDir), but how it can solve existing problems?

Reply:
@evsh 
**Small Summary:**
We have two main problems related to category names:
1. Empty name means "No category". So we can use one of two approaches: either don't allow to create real TorrentCategory instance with empty name (as it is carrently) or create this instance first time and consider it as "Default" category (use it instead of "No category" term) but we should prevent its deletion/editing by user.
2. "Unclean" names. Before this PR we allowed only valid names, but the users requested "transparent correction" of such "unclean" names (eg. name with backslashes instead of slashes in it). This "correction" leads to emergence of some sets of "source" category names which map into the same "clean" name. It requires to have some "non-kosher" code (you don't like here), or look for another solution which I have not seen yet.

Reply:
If we take the Composite pattern, the root object will be bound to the default save path (and, in general, to the default settings). Seems like the Category class has to have "bool editable()" property if you want to strictly follow the current behaviour, but I would simply replace the default settings with the root category and edit default save path via editing the root category object (subclass of the Category class) instead. 

As to the second item, I don't really understand the problem, unfortunately. If torrent → category mapping occurs only when adding a torrent or editing its properties, does it mean that category name correction is required only at that times? And after that one can establish the mapping by keeping a pointer to category object in the TorrentHandle. Between link via name and link via pointer I would choose the latter one because category removing seems to be exceptional case for me, category renaming is transparent, and category property change is explicit user action and as such we can collect all the affected torrents at that time.

Reply:
I thought about this approach earlier (do you not think that I have implemented the first come to mind?). But some things have put me off it. Don't know, maybe I'm just not gone all the way to the end?..
>As to the second item, I don't really understand the problem, unfortunately. If torrent → category mapping occurs only when adding a torrent or editing its properties, does it mean that category name correction is required only at that times? And after that one can establish the mapping by keeping a pointer to category object in the TorrentHandle.

Category name correction is required everytime we need access category by its name:
1. When we create category (both when adding new category or loading categories from config file).
2. When we need any access to category (get/set properties, assign category to torrent, remove category) via some kind of serial API (e.g. WebAPI).
So I don't know how to do it only inside the TorrentCategory class, without having "category name cleaner" outside of it.

The one more question is how to store categories tree in config:
1. The current "flat" approach, although it does not reflect the imagery of the tree structure, at the same time ideal for storing the Categories in the "Subcategories" mode and without it (besides, it allows the "unclean" names).
2. Treelike approach can be also used for both "Subcategories" and "Flat categories" modes, but it not protected from the presence of inconsistent data. So how should we resolve such conflicts?
(The "store" problems above stems from the fact that the configuration files are user-editable otherwise they wouldn't exist)

Reply:
Category name correction is also required when we rename category (currently it's not possible). In fact, it turns out that if we want to enforce uniqueness of names, the Name is not an own property of the Category, although we are still convenient to access it as property.

Reply:
@evsh, I'm going to completely rewrite this PR, using a hierarchical model (Composite Pattern) and refusing certain assumptions that are no longer required (this also allows to avoid some potential problems that I described above, so that you can comment on them).
One thing I still want you to comment here is "Unique Names" issue. How would you implement it? Ie you need to have some list of objects with unique names (in addition to uniqueness, these names should have some constraints). Who and how should handle it? As I said above:
>In fact, it turns out that if we want to enforce uniqueness of names, the Name is not an own property of the Category, although we are still convenient to access it as property.